### PR TITLE
Improve comments in MapWnd.cpp GetPathsThroughSupplyLanes

### PR DESCRIPTION
**Pull request by LGM-Doyle**
If these comments are not sufficient or not clear then I'll try to improve them if you provide some feedback so I have direction on where I've been unclear.

### CODE REVIEW

File: UI/MapWnd.cpp [hunk 1]
```diff
@@ -2786,11 +2786,15 @@ namespace {
     // corresponding to valid supply lane destinations
     typedef boost::unordered_multimap<int,int> SupplyLaneMMap;
 
-    // AncestorInfo stores the \p ids of systems one back on the path.
+    // AncestorInfo stores the \p ids of systems one hop back on the path
+    // and the \p o originating terminal point that the path to this
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: there must be some way to make that more understandable...

File: UI/MapWnd.cpp [hunk 2]
```diff
@@ -2786,11 +2786,15 @@ namespace {
...
+    // ancestor started from.
     struct AncestorInfo {
-        AncestorInfo(int a, int o) : ids(1, a), single_origin(o) {}
-        std::vector<int> ids;  // previous systems
-        // originating terminal or nothing for connected resources
+        AncestorInfo(int a, int o) : one_hop_back(1, a), single_origin(o) {}
+        // systems one hop back on the path
+        std::vector<int> one_hop_back;
+        // originating terminal or none for resources connected to at
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: using commas would help a lot...

File: UI/MapWnd.cpp [hunk 3]
```diff
@@ -2805,6 +2809,44 @@ namespace {
        reachable \p terminal_points along the \p supply_lanes.
        b) empty indicating no more work to do.
 
+       Algorithm Descrition:
+
+       The algorithm starts with terminal points, which are system ids
+       and supply lanes which are pairs of system ids.  It finds all
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: (terminal points) are (system ids and supply lanes) ?

File: UI/MapWnd.cpp [hunk 4]
```diff
@@ -2805,6 +2809,44 @@ namespace {
...
+       paths from any terminal point to any other terminal point
+       connected only by supply lanes.
+
+       The algorithm has two parts:  find all paths between terminal
+       points along supply lanes and return a set of all the system ids
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: where is the separation between the parts in this sentence?

File: UI/MapWnd.cpp [hunk 5]
```diff
@@ -2805,6 +2809,44 @@ namespace {
...
+       When two paths meet from different terminal points it records
+       both points that met as mid points on a good path between
+       terminal points.
+
+       When two paths meet from the same terminal points it merges them
+       into one path.
+
+
+       In the second part, it starts from the mid points and works its
+       way back to the terminal points, recoding every system along the
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: recording?

File: UI/MapWnd.cpp [hunk 6]
```diff
@@ -2817,14 +2859,22 @@ namespace {
...
-        // A list of all reachable midpoints between at least two different terminal points
+        // ancestors holds systems already reached by the breadth first search.
+        boost::unordered_map<int, AncestorInfo> ancestors;
+
+        // reachable_midpoints holds all systems reachable from at least
+        // two different terminal points.
         std::vector<int> reachable_midpoints;
 
-        // Start from all terminal points.
+        // Initialize with all the terminal points, by addind all
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: adding

File: UI/MapWnd.cpp [hunk 7]
```diff
@@ -2817,14 +2859,22 @@ namespace {
...
+        // terminal points to the queue and to ancestors.
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: , and to

File: UI/MapWnd.cpp [hunk 8]
```diff
@@ -2835,73 +2885,105 @@ namespace {
         // Find all reachable midpoints where paths from two different
         // terminal points meet.
         while (!try_next.empty() ) {
+            // Try the next system from the queue.
             const PrevCurrInfo & curr = try_next.front();
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: no space before &

File: UI/MapWnd.cpp [hunk 9]
```diff
@@ -2835,73 +2885,105 @@ namespace {
...
+        // good paths between terminal points.
+
         // No terminal point has a path to any other terminal point.
         if (reachable_midpoints.empty())
             return;
 
         // Return all systems on any path back to a terminal point.
+        // Start from every mid point and back track along all paths
+        // from that mid point adding each system to the good path.
+        // Stop back tracking when you hit a system already on the good
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp:2988]: meaning of "good path" is unclear

File: UI/MapWnd.cpp [hunk 10]
```diff
@@ -2835,73 +2885,105 @@ namespace {
...
+        // All ancestors on the path(s) from this midpoint not yet tried.
         boost::unordered_set<int> ancestors_on_path;
+
         for (std::vector<int>::const_iterator reached_it = reachable_midpoints.begin();
              reached_it != reachable_midpoints.end(); ++reached_it)
         {
             boost::unordered_map<int, AncestorInfo>::const_iterator ancestor_ii_sys;
             int ii_sys;
+
+            // Add the mid point to ancestors_on_path and while there
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: comma in here please...

**Comment by LGM-Doyle** [comment]
I've tried to address everything.

I moved the introduction to before the utility classes, so that the purpose of the utility classes might be evident from the earlier description.  

I expanded the definition of good_path in the introduction.

File: UI/MapWnd.cpp [hunk 11]
```diff
@@ -2781,16 +2781,91 @@ void MapWnd::ClearSystemRenderingBuffers() {
...
+       \p supply_lanes are pairs of system ids at the end of supply
+       lanes.
+
+       GetPathsThroughSupplyLanes returns a \p good_path.
+
+       The \p good_path is all system ids of systems connecting any \p
+       terminal_point to any other reachable \p terminal_point along the
+       \p supply_lanes. The \p good_path is all systems that could
+       transport a resource from a source to a sink along a starlane
+       that is part of the supply. The \p good_path includes the
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: What does "that is part of the supply" mean?

File: UI/MapWnd.cpp [hunk 12]
```diff
@@ -2781,16 +2781,91 @@ void MapWnd::ClearSystemRenderingBuffers() {
...
+       terminal point system ids that are part of the network.  The \p
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: What is "the network"? Different from other terms for groups of connected systems used above?

File: UI/MapWnd.cpp [hunk 13]
```diff
@@ -2781,16 +2781,91 @@ void MapWnd::ClearSystemRenderingBuffers() {
...
+
+       Algorithm Descrition:
+
+       The algorithm starts with terminal points and supply lanes.  It
+       finds all paths from any terminal point to any other terminal
+       point connected only by supply lanes.
+
+       The algorithm has two parts:
+       1) find all paths between terminal points along supply lanes
+       2) return a set of all the system ids of the found paths.
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: What's the difference between finding and returning? Why does this make "two parts" of an algorithm and not two return values...?

File: UI/MapWnd.cpp [hunk 14]
```diff
@@ -2781,16 +2781,91 @@ void MapWnd::ClearSystemRenderingBuffers() {
...
+
+
+       In the first part, it starts a breadth first search from every
+       terminal point at once.  It tracks which terminal point each path
+       started from.
+
+       When two paths meet from different terminal points it records
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: Rephrase to "When two paths from different terminal points meet..."

File: UI/MapWnd.cpp [hunk 15]
```diff
@@ -2781,16 +2781,91 @@ void MapWnd::ClearSystemRenderingBuffers() {
...
+       part visits any system more than once.
+
+       The first part uses visited to track already visited systems.
+
+       The second part stops back tracking along paths when it reaches
+       systems already on the good path.
+
+     */
+    void GetPathsThroughSupplyLanes(
+        boost::unordered_set<int> & good_path,
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: remove spaces before &

File: UI/MapWnd.cpp [hunk 16]
```diff
@@ -2799,13 +2874,6 @@ namespace {
...
 
-    /**
-       Returns either:
-       a) a \p good_path between from any \p terminal_points to all
-       reachable \p terminal_points along the \p supply_lanes.
-       b) empty indicating no more work to do.
-
-     */
     void GetPathsThroughSupplyLanes(
         boost::unordered_set<int> & good_path,
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp:2881]: -space before &

File: UI/MapWnd.cpp [hunk 17]
```diff
@@ -2817,97 +2885,139 @@ namespace {
...
+            // in the good_path.
+            unprocessed.insert(*reached_it);
+            while (!unprocessed.empty()) {
+                ii_sys = *unprocessed.begin();
+                unprocessed.erase(unprocessed.begin());
+
+                // If ii_sys is not in the good_path, then add it to the
+                // good_path and add all of its visited to the unprocessed.
+                if ((previous_ii_sys = visited.find(ii_sys)) != visited.end()
+                    && (good_path.count(ii_sys) == 0 ))
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: extra space after 0 ?

**Comment by LGM-Doyle** [comment]
Thanks for the help clarifying the comments.

**Pull request merged**